dll/win32/shell32/CUserNotification.cpp

```diff
@@ -0,0 +1,637 @@
+
+#include <mmsystem.h>
+#undef PlaySound
+
+WINE_DEFAULT_DEBUG_CHANNEL(shell_notify);
+
+
+/* Use Windows-compatible window callback message */
+#define WM_TRAYNOTIFY   (WM_USER + 100)
+
+/* Notification icon ID */
+#define ID_NOTIFY_ICON  0
+
+/* Balloon timers */
+#define ID_BALLOON_TIMEOUT      1
+#define ID_BALLOON_DELAYREMOVE  2
+#define ID_BALLOON_QUERYCONT    3
+#define ID_BALLOON_SHOWTIME     4
+
+#define BALLOON_DELAYREMOVE_TIMEOUT 250 // milliseconds
+
+
+CUserNotification::CUserNotification() :
+    m_hWorkerWnd(NULL),
+    m_hIcon(NULL),
+    m_uTimeout(15000),
+    m_uInterval(10000),
+    m_uContinuePoolInterval(0),
+    m_cRetryCount(0),
+    m_dwInfoFlags(0),
+    m_hRes(S_OK),
+    m_pqc(NULL)
+{
+}
+
+CUserNotification::~CUserNotification()
+{
+    /* If we have a notification window... */
+    if (m_hWorkerWnd)
+    {
+        /* ... remove the notification icon and destroy the window */
+        RemoveIcon();
+        ::DestroyWindow(m_hWorkerWnd);
+        m_hWorkerWnd = NULL;
+    }
+
+    /* Destroy our local icon copy */
+    if (m_hIcon)
+        ::DestroyIcon(m_hIcon);
+}
+
+VOID CUserNotification::RemoveIcon()
+{
+    NOTIFYICONDATAW nid = {0};
+
+    nid.cbSize = NOTIFYICONDATAW_V3_SIZE; // sizeof(nid);
+    nid.hWnd = m_hWorkerWnd;
+    nid.uID  = ID_NOTIFY_ICON;
+
+    /* Remove the notification icon */
+    ::Shell_NotifyIconW(NIM_DELETE, &nid);
+}
+
+VOID CUserNotification::DelayRemoveIcon(IN HRESULT hRes)
+{
+    /* Set the return value for CUserNotification::Show() and defer removing icon */
+    m_hRes = hRes;
+    ::SetTimer(m_hWorkerWnd, ID_BALLOON_DELAYREMOVE,
+               BALLOON_DELAYREMOVE_TIMEOUT, NULL);
+}
+
+VOID CUserNotification::TimeoutIcon()
+{
+    /*
+     * The balloon timed out, we need to wait before showing it again.
+     * If we retried too many times, delete the notification icon.
+     */
+    if (m_cRetryCount > 0)
+    {
+        /* Decrement the retry count */
+        --m_cRetryCount;
+
+        /* Set the timeout interval timer */
+        ::SetTimer(m_hWorkerWnd, ID_BALLOON_TIMEOUT, m_uInterval, NULL);
+    }
+    else
+    {
+        /* No other retry: delete the notification icon */
+        DelayRemoveIcon(HRESULT_FROM_WIN32(ERROR_CANCELLED));
+    }
+}
+
+VOID CUserNotification::SetUpNotifyData(
+    IN UINT uFlags,
+    IN OUT PNOTIFYICONDATAW pnid)
+{
+    pnid->cbSize = NOTIFYICONDATAW_V3_SIZE; // sizeof(nid);
+    pnid->hWnd = m_hWorkerWnd;
+    pnid->uID  = ID_NOTIFY_ICON;
+    // pnid->uVersion = NOTIFYICON_VERSION;
+
+    if (uFlags & NIF_MESSAGE)
+    {
+        pnid->uFlags |= NIF_MESSAGE;
+        pnid->uCallbackMessage = WM_TRAYNOTIFY;
+    }
+
+    if (uFlags & (NIF_ICON | NIF_TIP))
+    {
+        pnid->uFlags |= NIF_ICON | NIF_TIP;
+
+        if (m_hIcon)
+            pnid->hIcon = m_hIcon;
+        else
+            pnid->hIcon = LoadIcon(NULL, IDI_WINLOGO);
+
+        ::StringCchCopyW(pnid->szTip, _countof(pnid->szTip), m_szTip.GetString());
```

Comment:
If only NIF_ICON is set you still specify a tip?

Reply:
This function is a private helper function, which is called internally in SetIconInfo() (and in Show()). The uFlags parameter of this function indicates which information I want to set/update from where. When I pass NIF_ICON in the uFlags when calling this helper function, it's actually because I'm going to update the icon and the tips: just imagine actually that I'm putting all this block of code in SetIconInfo() and in Show().

Reply:
I improved this function a bit following your comment.